### PR TITLE
Enhancement: Integration of Prometheus exporter.

### LINKED ISSUE #1558: Enhancement: Prometheus exporter for Litestar
### Summary

It looks lite the existing solutions like https://github.com/stephenhillier/starlette_exporter and https://github.com/trallnag/prometheus-fastapi-instrumentator don't work with Litestar as they rely on Starlette`s add_middleware() method.
Litestar does have OpenTelemetry integration, but as I understand it is not compatible with Prometheus as it cannot expose /metrics endpoint.

### Basic Example

_No response_

### Drawbacks and Impact

_No response_

### Unresolved questions

_No response_

<a href="https://polar.sh/litestar-org/litestar/issues/1558">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/1558/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by r4gesingh47**
1. Integrated Prometheus middleware, which record four golden signals metrics.

2. Integrate the prometheus exporter controller which also supports openmetrics format.

Closes #1558

### CODE REVIEW

**Comment by peterschutt** [comment]
Thanks for the PR @r4gesingh47 - looks like you might need to run `poetry lock`?

File: pyproject.toml [hunk 1]
```diff
@@ -93,6 +93,7 @@ structlog = { version = "*", optional = true }
 typing-extensions = "*"
 uvicorn = {extras = ["standard"], version = "^0.21.1", optional = true}
 importlib-resources = { version="^5.12.0", python = "<3.9" }
+prometheus-client = "^0.16.0"
```
Comment (cofin) [inline pyproject.toml:96]: This should be an optional dependency.  I think it's probably best to create a `prometheus` group or something similar to install the additional dependency.

**Comment by cofin** [comment]
Nice work.  This will be a welcomed addition.  I think it may be worthwhile to utilize the [InitPluginProtocol](https://github.com/litestar-org/litestar/blob/main/litestar/plugins.py#L24) class to configure everything at startup with a single class.

**Review by JacobCoffee — COMMENTED** [review]
Looks good! Thank you for your work.. ill be testing this out in my lab soon. 

+1 on making the prom client package optional

**Comment by r4gesingh47** [comment]
> Thanks for the PR @r4gesingh47 - looks like you might need to run `poetry lock`?

@peterschutt  running poetry lock is showing unnecessary changes of files of other dependencies.

**Comment by provinzkraut** [comment]
> running poetry lock is showing unnecessary changes of files of other dependencies.

You need to run `poetry lock --no-upadte`

File: pyproject.toml (hunk 1 above)
Comment (r4gesingh47) [reply pyproject.toml:96]: > This should be an optional dependency. I think it's probably best to create a `prometheus` group or something similar to install the additional dependency.

Hi @cofin sorry did not realized that, Is it ok to add it in opentelemetry group in tools.poetry.extras or it is best to create a new one.

Comment (cofin) [reply pyproject.toml:96]: I would create a new one.  This will make it a totally optional component that can be installed without OTEL.

**Review by Goldziher — COMMENTED** [review]
looks very nice! thanks for contributing 😉 

Please add docs and examples.

File: litestar/contrib/prometheus/middleware.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+from litestar.exceptions import MissingDependencyException
+from litestar.middleware.base import AbstractMiddleware
+
+__all__ = ("PrometheusMiddleware",)
+
+
+try:
+    import prometheus_client  # noqa: F401
+except ImportError as e:
+    raise MissingDependencyException("prometheus_client") from e
```
[the remark this answers is not in the document]
Comment (r4gesingh47) [reply litestar/contrib/prometheus/middleware.py:18]: I did not the point, rest everywhere I see it is mentioned in this format only.

File: litestar/contrib/prometheus/controller.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+__all__ = [
+    "PrometheusController",
+]
+
+
+class PrometheusController(Controller):
+    """Controller for Prometheus endpoints.
+    Example usage:
```
[the remark this answers is not in the document]
Comment (r4gesingh47) [reply litestar/contrib/prometheus/controller.py:35]: Ok will add it surely, might be a little late running busy, max 2 days.

File: litestar/contrib/prometheus/middleware.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        """Middleware that adds Prometheus instrumentation to the application.
+
+        Args:
+            app: The ``next`` ASGI app to call.
+            config: An instance of :class:`PrometheusConfig <.contrib.prometheus.PrometheusConfig>`
+        """
+        super().__init__(app=app, scopes=config.scopes, exclude=config.exclude, exclude_opt_key=config.exclude_opt_key)
+        self.config = config
+        self.labels = OrderedDict(self.config.labels) if self.config.labels is not None else None
+        self.kwargs = {}
```
[the remark this answers is not in the document]
Comment (r4gesingh47) [reply litestar/contrib/prometheus/middleware.py:44]: Done, will push it with docs changes only.

File: litestar/contrib/prometheus/controller.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import os
+
+from litestar import Controller, get
+from litestar.exceptions import MissingDependencyException
+from litestar.response import Response
+
+try:
+    import prometheus_client  # noqa: F401
+except ImportError as e:
+    raise MissingDependencyException("prometheus_client") from e
```
[the remark this answers is not in the document]
Comment (provinzkraut) [reply litestar/contrib/prometheus/controller.py:12]: This is the correct format

File: litestar/contrib/prometheus/middleware.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (provinzkraut) [reply litestar/contrib/prometheus/middleware.py:18]: @r4gesingh47 the format is correct here, not action needs to be taken.

**Comment by Goldziher** [comment]
hi, where are you standing with this @r4gesingh47 ?

**Comment by r4gesingh47** [comment]
> hi, where are you standing with this @r4gesingh47 ?

Hi, will be pushing in this week, done with examples just have to write the testcases.

**Comment by r4gesingh47** [comment]
> hi, where are you standing with this @r4gesingh47 ?

Hey added the examples and tests for that.

**Comment by Goldziher** [comment]
> > hi, where are you standing with this @r4gesingh47 ?
> 
> Hey added the examples and tests for that.

great, please resolve the merge conflicts so we can proceed.

**Comment by r4gesingh47** [comment]
@Goldziher Hi, can you help me why the above tests are failing, it is due to some import issue I guess but was not happening earlier

**Comment by Goldziher** [comment]
> @Goldziher Hi, can you help me why the above tests are failing, it is due to some import issue I guess but was not happening earlier

seems a test is missing?

**Comment by r4gesingh47** [comment]
> > @Goldziher Hi, can you help me why the above tests are failing, it is due to some import issue I guess but was not happening earlier
> 
> seems a test is missing?

I don't think so, specifc error is ValidPython name for a module when i am running locally just for prometheus it is working fine.

**Comment by Goldziher** [comment]
> > > @Goldziher Hi, can you help me why the above tests are failing, it is due to some import issue I guess but was not happening earlier
> > 
> > 
> > seems a test is missing?
> 
> I don't think so, specifc error is ValidPython name for a module when i am running locally just for prometheus it is working fine.

can i push to your branch?

**Comment by r4gesingh47** [comment]
> > > > @Goldziher Hi, can you help me why the above tests are failing, it is due to some import issue I guess but was not happening earlier
> > > 
> > > 
> > > seems a test is missing?
> > 
> > 
> > I don't think so, specifc error is ValidPython name for a module when i am running locally just for prometheus it is working fine.
> 
> can i push to your branch?

What was the issue?

**Pull request merged**